### ISSUE TITLE
placeholder appears in result

### ISSUE THREAD

**Issue by samth** [racket/lazy#3]
This program, due to @jasonhemann, produces the output `'((constant) . #<placeholder>)`.

```
#lang lazy
(require lazy/force)

(define (pull t$) (if (eq? (car t$) 'tag1) (pull (cdr t$)) t$))
(define (relρ a) (cons 'tag1 (cons 'tag2 (cons a (relρ a)))))
;; change either a in the body to 'const 

(!! 
 (let (($ (cdr (pull (relρ '(constant)))))) ;; s/'(constant)/'constant/
   (cons (car $) (cadr (pull (cdr $)))))) ;; change (car $) to 'other-constant
      ;; or replace the above with (cadr (pull (cdr $)))
      ;; or in-line the let in both places
      ;; also interesting, swap the order of args to cons
```

**Comment by samth** [comment]
cc @stchang

**Comment by elibarzilay** [comment]
*sigh* at the forgotten art of minimizing bugs...  A much simpler test:

    (!! (let ([l '(x)]) (list l l)))

And *sigh* for bad code...  There's a trick in lazy/promise that is supposed to reduce the number of placeholders, which AFAICT is just broken.  Removing this trick seems to work (but someone should go over this more seriously) -- replace the 12-line `(let ([y (hash-ref t x #f)]) ...)` expression on line 58 with:

        (hash-ref t x
          (λ() (set! placeholders? #t)
               (define p (make-placeholder #f))
               (hash-set! t x p)
               (placeholder-set! p expr)
               (placeholder-get p)))

(And the comment before that thing should be updated too.)

Here's a way to see that it works (I'm not sure if this is new or not, but without the `~s` printing you don't see the graph):

    #lang lazy
    (require "force.rkt")
    
    (! (print-graph #t))
    
    (! (eprintf "~s\n" (!! (letrec ([l (cons 1 l)]) l))))
    (! (eprintf "~s\n" (!! (let ([l '(x)]) (list l l l l)))))

**Comment by spdegabrielle** [comment]
is this still an issue?  Please consider adding the label `good first issue` so it is findable: 

[Issues labeled `good first issue` in in Racket GitHub repositories](https://github.com/search?q=type:issue+org:racket+state:open+label:%22good+first+issue%22)

**Comment by jasonhemann** [comment]
Yes, this issue is still outstanding. Using @elibarzilay 's simplified test case.

```
#lang lazy
(!! (let ([l '(x)]) (list l l)))
```

```
; 
; Welcome to Racket v7.7.0.4.
; 
> 
'((x) #<placeholder>)
```

**Comment by jasonhemann** [comment]
Perhaps a member of the Racket organization could add that label?
